### PR TITLE
Added SACD implementation.

### PR DESCRIPTION
**Pull request by PKWadsy**
Tested on Cart-Pole. Attains maximum reward

based on [this](https://arxiv.org/pdf/1910.07207) paper

### CODE REVIEW

File: cares_reinforcement_learning/algorithm/policy/SACD.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        gamma: float,
+        tau: float,
+        reward_scale: float,
+        action_num: int,
+        actor_lr: float,
+        critic_lr: float,
+        alpha_lr: float,
+        device: torch.device,
+    ):
+        self.type = "discrete_policy"
```
Comment (beardyFace) [inline cares_reinforcement_learning/algorithm/policy/SACD.py:34]: How does this get handled by gym_environment? is there a respective update? This should just stay as policy no?

File: cares_reinforcement_learning/networks/SACD/actor.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(
+        self,
+        observation_size: int,
+        num_actions: int,
+        hidden_size: list[int] = None,
+        log_std_bounds: list[int] = None,
+    ):
+        super().__init__()
+        if hidden_size is None:
+            hidden_size = [256, 256]
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/SACD/actor.py:20]: double check default hidden layer size

File: cares_reinforcement_learning/networks/SACD/actor.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        if log_std_bounds is None:
+            log_std_bounds = [-20, 2]
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/SACD/actor.py:22]: double check log_std_bounds default

File: cares_reinforcement_learning/networks/SACD/actor.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    def forward(
+        self, state: torch.Tensor
+    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        action_probs = self.act_net(state)
+        max_probability_action = torch.argmax(action_probs)
+        dist = torch.distributions.Categorical(action_probs)
+        action = dist.sample()
+        # Offset any values which are zero by a small amount so no nan nonsense
+        zero_offset = action_probs == 0.0
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/SACD/actor.py:44]: to avoid confusion just through () around the (action_probs == 0.0)

File: cares_reinforcement_learning/networks/SACD/critic.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+class Critic(nn.Module):
+    def __init__(
+        self,
+        observation_size: int,
+        num_actions: int,
+        hidden_size: list[int] = None,
+    ):
+        super().__init__()
+        if hidden_size is None:
+            hidden_size = [256, 256]
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/SACD/critic.py:14]: double check hidden layer size defaults

File: cares_reinforcement_learning/util/configurations.py [hunk 6]
```diff
@@ -209,6 +209,17 @@ class SACAEConfig(AlgorithmConfig):
...
 
+class SACDConfig(AlgorithmConfig):
+    algorithm: str = Field("SACD", Literal=True)
+    actor_lr: Optional[float] = 3e-4
+    critic_lr: Optional[float] = 3e-4
+    alpha_lr: Optional[float] = 3e-4
+
+    gamma: Optional[float] = 0.99
+    tau: Optional[float] = 0.005
+    reward_scale: Optional[float] = 1.0
```
Comment (beardyFace) [inline cares_reinforcement_learning/util/configurations.py:220]: double check defaults from paper/code

**Review by beardyFace — CHANGES_REQUESTED** [review]

File: cares_reinforcement_learning/algorithm/policy/SACD.py (hunk 1 above)
Comment (PKWadsy) [reply cares_reinforcement_learning/algorithm/policy/SACD.py:34]: The normalization in the gym environment was breaking things so wanted to make a loop without normalization. Sorry about the horrible naming

File: cares_reinforcement_learning/networks/SACD/critic.py (hunk 5 above)
Comment (PKWadsy) [reply cares_reinforcement_learning/networks/SACD/critic.py:14]: Where can we get this info? The paper?

File: cares_reinforcement_learning/algorithm/policy/SACD.py (hunk 1 above)
Comment (beardyFace) [reply cares_reinforcement_learning/algorithm/policy/SACD.py:34]: how is it breaking things? it shouldn't break anything

File: cares_reinforcement_learning/networks/SACD/critic.py (hunk 5 above)
Comment (beardyFace) [reply cares_reinforcement_learning/networks/SACD/critic.py:14]: The paper or the code they provided

File: cares_reinforcement_learning/networks/SACD/actor.py (hunk 3 above)
Comment (beardyFace) [reply cares_reinforcement_learning/networks/SACD/actor.py:22]: can remove log_std_bounds not even used here

File: cares_reinforcement_learning/networks/SACD/actor.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import torch
+from torch import nn
+
+from cares_reinforcement_learning.util.common import SquashedNormal
```
Comment (beardyFace) [inline cares_reinforcement_learning/networks/SACD/actor.py:4]: can also remove the squashednormal import here

**Review by beardyFace — CHANGES_REQUESTED** [review]

File: cares_reinforcement_learning/algorithm/policy/SACD.py (hunk 1 above)
Comment (SamBoasman) [reply cares_reinforcement_learning/algorithm/policy/SACD.py:35]: In the discrete setting, `select_action_from_policy()` and `sample_action()` return a tensor containing the integer index of the action to enact. However, the action returned from `sample_action()` is normalised, while the action returned from `select_action_from_policy()` is denormalised. To make this work, we would need to configure some sort of redundant reversal of normalisation/denormalisation for one of these functions so that run_action_on_emulator() receives only the normalised or original action format instead of both.

**Review by beardyFace — APPROVED** [review]
